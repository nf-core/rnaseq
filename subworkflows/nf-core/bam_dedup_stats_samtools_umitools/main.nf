//
// UMI-tools dedup, index BAM file and run samtools stats, flagstat and idxstats
//

include { UMITOOLS_DEDUP     } from '../../../modules/nf-core/umitools/dedup/main'
include { SAMTOOLS_INDEX     } from '../../../modules/nf-core/samtools/index/main'
include { BAM_STATS_SAMTOOLS } from '../bam_stats_samtools/main'

workflow BAM_DEDUP_STATS_SAMTOOLS_UMITOOLS {
    take:
    ch_bam_bai          // channel: [ val(meta), path(bam), path(bai/csi) ]
    val_get_dedup_stats // boolean: true/false

    main:

    //
    // UMI-tools dedup
    //
    UMITOOLS_DEDUP ( ch_bam_bai, val_get_dedup_stats )

    //
    // Index BAM file and run samtools stats, flagstat and idxstats
    //
    SAMTOOLS_INDEX ( UMITOOLS_DEDUP.out.bam )

    ch_bam_bai_dedup = UMITOOLS_DEDUP.out.bam
        .join(SAMTOOLS_INDEX.out.bai, by: [0], remainder: true)
        .join(SAMTOOLS_INDEX.out.csi, by: [0], remainder: true)
        .map {
            meta, bam, bai, csi ->
                if (bai) {
                    [ meta, bam, bai ]
                } else {
                    [ meta, bam, csi ]
                }
        }

    BAM_STATS_SAMTOOLS ( ch_bam_bai_dedup, [ [:], [] ] )

    emit:
    bam      = UMITOOLS_DEDUP.out.bam          // channel: [ val(meta), path(bam) ]
    deduplog = UMITOOLS_DEDUP.out.log          // channel: [ val(meta), path(log) ]

    bai      = SAMTOOLS_INDEX.out.bai          // channel: [ val(meta), path(bai) ]
    csi      = SAMTOOLS_INDEX.out.csi          // channel: [ val(meta), path(csi) ]
    stats    = BAM_STATS_SAMTOOLS.out.stats    // channel: [ val(meta), path(stats) ]
    flagstat = BAM_STATS_SAMTOOLS.out.flagstat // channel: [ val(meta), path(flagstat) ]
    idxstats = BAM_STATS_SAMTOOLS.out.idxstats // channel: [ val(meta), path(idxstats) ]

}
