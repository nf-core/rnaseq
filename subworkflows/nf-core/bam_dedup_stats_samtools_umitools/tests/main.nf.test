nextflow_workflow {

    name "Test Workflow BAM_DEDUP_STATS_SAMTOOLS_UMITOOLS"
    script "../main.nf"
    workflow "BAM_DEDUP_STATS_SAMTOOLS_UMITOOLS"
    tag "subworkflows"
    tag "subworkflows_nfcore"
    tag "subworkflows/bam_dedup_stats_samtools_umitools"
    tag "subworkflows/bam_stats_samtools"
    tag "bam_dedup_stats_samtools_umitools"
    tag "bam_stats_samtools"
    tag "samtools"
    tag "samtools/index"
    tag "samtools/stats"
    tag "samtools/idxstats"
    tag "samtools/flagstat"
    tag "umitools"
    tag "umitools/dedup"

    test("sarscov2_bam_bai") {

        when {
            params {
                outdir   = "$outputDir"
            }
            workflow {
                """
                val_get_dedup_stats = false

                input[0] = Channel.of([
                    [ id:'test'], // meta map
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/bam/test.paired_end.sorted.bam', checkIfExists: true),
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/bam/test.paired_end.sorted.bam.bai', checkIfExists: true)
                ])
                input[1] = val_get_dedup_stats
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert workflow.out.bam.get(0).get(1) ==~ ".*.bam"},
                { assert workflow.out.bai.get(0).get(1) ==~ ".*.bai"},
                { assert snapshot(workflow.out.stats).match("test_bam_dedup_stats_samtools_umitools_stats") },
                { assert snapshot(workflow.out.flagstat).match("test_bam_dedup_stats_samtools_umitools_flagstats") },
                { assert snapshot(workflow.out.idxstats).match("test_bam_dedup_stats_samtools_umitools_idxstats") }
            )
        }

    }

}
