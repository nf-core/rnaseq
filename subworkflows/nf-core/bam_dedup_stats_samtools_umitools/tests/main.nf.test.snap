{
    "test_bam_dedup_stats_samtools_umitools_idxstats": {
        "content": [
            [
                [
                    {
                        "id": "test"
                    },
                    "test.idxstats:md5,1adb27b52d4d64b826f48b59d61dcd4d"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.01.0"
        },
        "timestamp": "2023-11-06T09:58:40.394333937"
    },
    "test_bam_dedup_stats_samtools_umitools_flagstats": {
        "content": [
            [
                [
                    {
                        "id": "test"
                    },
                    "test.flagstat:md5,0bb716e40fae381b97484b58e0b16efe"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.01.0"
        },
        "timestamp": "2023-11-06T09:58:40.385185447"
    },
    "test_bam_dedup_stats_samtools_umitools_stats": {
        "content": [
            [
                [
                    {
                        "id": "test"
                    },
                    "test.stats:md5,02342d307779941001376ff5d19e941a"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.01.0"
        },
        "timestamp": "2024-02-13T16:45:33.563016651"
    }
}